ENIGMAsystem/SHELL/Graphics_Systems/Direct3D11/DX11vertex.cpp

```diff
@@ -16,23 +16,274 @@
+    // create the pixel shader
+    graphics_compile_shader(g_strPS, &pBlobPS, "PS", "PS", "ps_4_0");
+    m_device->CreatePixelShader(pBlobPS->GetBufferPointer(), pBlobPS->GetBufferSize(),
+                                NULL, &g_pPixelShader);
+  }
+
+  m_deviceContext->VSSetShader(g_pVertexShader, NULL, 0);
+  m_deviceContext->PSSetShader(g_pPixelShader, NULL, 0);
+}
 
 }
 
 namespace enigma_user {
 
-void vertex_argb(int buffer, unsigned argb) {}
+void vertex_argb(int buffer, unsigned argb) {
+  enigma::vertexBuffers[buffer]->vertices.push_back(argb);
+}
 
-void vertex_color(int buffer, int color, double alpha) {}
+void vertex_color(int buffer, int color, double alpha) {
+  enigma::color_t finalcol = (CLAMP_ALPHA(alpha) << 24) | (COL_GET_R(color) << 16) | (COL_GET_G(color) << 8) | COL_GET_B(color);
+  enigma::vertexBuffers[buffer]->vertices.push_back(finalcol);
+}
 
-void vertex_submit(int buffer, int primitive, unsigned vertex_start, unsigned vertex_count) {}
+void vertex_submit(int buffer, int primitive, unsigned start, unsigned count) {
+  const enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+  const enigma::VertexFormat* vertexFormat = enigma::vertexFormats[vertexBuffer->format];
 
-void index_submit(int buffer, int vertex, int primitive, unsigned start, unsigned count) {}
+  enigma::graphics_prepare_default_shader();
+  enigma::graphics_prepare_buffer(buffer, false);
+
+  UINT stride = 0;
+  ID3D11InputLayout* vertexLayout = vertex_format_layout(vertexFormat, stride);
+  m_deviceContext->IASetInputLayout(vertexLayout);
+
+  UINT offset = 0;
+  ID3D11Buffer* vertexBufferPeer = vertexBufferPeers[buffer];
+  m_deviceContext->IASetVertexBuffers(0, 1, &vertexBufferPeer, &stride, &offset);
+
+  m_deviceContext->IASetPrimitiveTopology(primitive_types[primitive]);
+  m_deviceContext->Draw(count, start);
+
+  vertexLayout->Release();
+}
+
+void index_submit(int buffer, int vertex, int primitive, unsigned start, unsigned count) {
+  const enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[vertex];
+  const enigma::VertexFormat* vertexFormat = enigma::vertexFormats[vertexBuffer->format];
+  const enigma::IndexBuffer* indexBuffer = enigma::indexBuffers[buffer];
+
+  enigma::graphics_prepare_default_shader();
+  enigma::graphics_prepare_buffer(buffer, true);
+  enigma::graphics_prepare_buffer(vertex, false);
+
+  UINT stride = 0;
+  ID3D11InputLayout* vertexLayout = vertex_format_layout(vertexFormat, stride);
+  m_deviceContext->IASetInputLayout(vertexLayout);
+
+  UINT offset = 0;
+  ID3D11Buffer* vertexBufferPeer = vertexBufferPeers[vertex];
+  m_deviceContext->IASetVertexBuffers(0, 1, &vertexBufferPeer, &stride, &offset);
+
+  ID3D11Buffer* indexBufferPeer = indexBufferPeers[buffer];
+  m_deviceContext->IASetIndexBuffer(
+    indexBufferPeer,
+    (indexBuffer->type == index_type_uint) ? DXGI_FORMAT_R32_UINT : DXGI_FORMAT_R16_UINT,
+    0
+  );
+
+  m_deviceContext->IASetPrimitiveTopology(primitive_types[primitive]);
```

Comment:
On what, the `primitive_types` array? Sure, I'll do it before I merge.

Reply:
Affirmative.

Reply:
Alright, I added a macro for it: 86c5f8751813a0b7bcae66dc446e60e6f0fe73d8

I'm unsure of it, so I will wait to hear back on this before applying it to the other systems and merging. I have the following questions.
1) I need to release `vertexLayout` for `vertex_submit` and `index_submit`, is it ok to do that in the macro?
2) In `primitive >= (int)primitive_types_size` the former is an `int` and the latter is a `size_t`. Which cast is better to avoid a warning? Should I cast `primitive_types_size` to `int` or `primitive` to `size_t`?

Reply:
Alright, I chose the least suckiest cast, and I am happy just having the macro release before the return. I may add caching to this later when I benchmark the vertex functions anyway, in which case I won't even need to release the vertex layout. This is good.